File: CN Lab/18501A0591_3.py
#18501A0591_3.py
class node:
    def __init__(self):
        self.predecessor = None
        self.length = None
        self.label = False #tentative

class graph:
    def __init__(self):
        self.adj = []
        self.n = None
        self.path = []
    
    def dijsktra(self, src, des):

        ele = [node() for i in range(self.n)]
        ele[src].length = 0
        ele[src].label  = True #permanent

        k = src
        while k!=des:
            for i in range(self.n):
                if self.adj[k][i]!=0 and not ele[i].label:
                    ele[i].predecessor = k
                    ele[i].length = ele[k].length+self.adj[k][i]
        
            k = 0
            min = 10**9
            for i in range(self.n):
                if (not ele[i].label) and ele[i].length<min:
                    min = ele[i].length
                    k = i
            ele[k].label = True #permanent
        
        i,k = 1,des
        while k:
            self.path.append(k)
            k = ele[k].predecessor
            i+=1
        self.path.append(src)
        print(self.path)
        
        return i
    
    def print(self, n, ip2):
        print('The shortest path is:')
        for i in range(n-1,-1,-1):
            print(chr(self.path[i]+97),end='')
            if i>0:
                print('-->',end='')
        print('\n\nThe fwd tables are:')
        print('Dest address\t\tO/p Interface')
        for i in range(n-1,-1,-1):
            print(ip2,chr(self.path[i]+97))


g = graph()
g.n = int(input('enter number of nodes:'))
print('enter adjacent matrix:')
for i in range(g.n):
    g.adj.append(list(map(int,input().split())))

src,des = input('Enter src and dest:').split()
ip1,ip2 = input("Enter Ip's of src & dest:").split()

ans = g.dijsktra(ord(src)-97,ord(des)-97)

g.print(ans,ip2)

File: CN Lab/18501A0591_3.c
/*dijkstra*/
#include<stdio.h>
#include<limits.h>
#define max 10
int n,adj_mat[max][max];
struct node{
    int predecessor;
    int length;
    enum {permanent,tentative} label;
};
int shortestPath(int source,int des,int path[]){
    int i,j,min,k;
    struct node ele[n];
    for(i=0;i<n;i++){
            ele[i].predecessor=-1;
    ele[i].length=INT_MAX;
    ele[i].label=tentative;
}
ele[source].length=0;
ele[source].label=permanent;
k=source;
do{
        for(i=0;i<n;i++){
            if((adj_mat[k][i]!=0)&&(ele[i].label==tentative)){
                if(adj_mat[k][i]+ele[k].length<ele[i].length){
                    ele[i].predecessor=k;
ele[i].length=ele[k].length+adj_mat[k][i];
}
}
}
k=0;
min=INT_MAX;
for(i=0;i<n;i++){
if(ele[i].label==tentative&&ele[i].length<min){
min=ele[i].length;
k=i;
}
}
ele[k].label=permanent;
}while(k!=des);
i=0;
k=des;
do{
path[i++]=k;
k=ele[k].predecessor;
}while(k>=source);
printf("Distance :%d\n",ele[des].length);
return i;
}
int main(){
    printf("Enter number of nodes..:\n");
    scanf("%d",&n);
    printf("Enter graph in adjacency matrix form..:\n");
    for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                scanf("%d",&adj_mat[i][j]);
    }
    }
char source,des;
printf("Enter Source and Destination..:\n");
scanf(" %c %c",&source,&des);
int result=97;
if(source>='A'&&source<='Z'){
        result=65;
}
char ip1[20],ip2[20];
printf("Enter IP Address of source and destination..:\n");
scanf("%s %s",ip1,ip2);
int path[max];
int ans=shortestPath(source-result,des-result,path);
printf("The Shortest Path is..:");
for(int i=ans-1;i>=0;i--){
if(i==0){
printf("%c\n",path[i]+result);
continue;
}
printf("%c-->",path[i]+result);
}
printf("The forwarding tables are..:\n");
printf("Destination Address\t\tOutput Interface\n");
for(int i=ans-1;i>=0;i--){
printf("%s\t\t\t%c\n",ip2,path[i]+result);
}

return 0;
}
